### PR TITLE
Add enabled, disabled, enable, disable to runit

### PR DESCRIPTION
**Pull request by markkimsal**
New methods test if the standard 'run' file in the services folder has
the executable bit set.  When the 'run' file is not executable, runit
will not launch the service on boot.  This mimicks the enabled/diabled
behavior of other service providers.

The file permissions are set to 700 and 600 for enabled and disabled
states.  Because of the ad-hoc nature that runit service files can
sometimes posess, the run files may have passwords or other sensitive
information.  Therefore, the permissions are hard coded to only be
readable by the file owner, who is also the user who runs the process.

This patch enables salt-ssh to use serivce.\* calls when ssh_minion_opts
sets runit as a global service provider.  It also allows adding "enabled" keys
to state files that manage services and may or may not override the
provider key

```
enable_service:
    service.running:
       - name: mysql
       - provider: {{ mysql.service_provider }}  <== overridable
       - enable: True     <== breaks because runit doesn't have enable
       - reload: True
```

### CODE REVIEW

File: salt/modules/runit.py [hunk 1]
```diff
@@ -27,6 +27,24 @@
...
+#    cmd = 'sv'
+#    which_result = __salt__['cmd.which_bin']([cmd])
+#    if which_result is None:
+#        raise CommandNotFoundError(
+#            'Could not find a `{0}` binary'.format(cmd)
+#        )
+#    #change __virtualnamne__
+#    return 'runit';
+#
+
```
Comment (basepi) [inline salt/modules/runit.py]: Should we remove these commented lines or uncomment them?

**Comment by basepi** [comment]
Left one comment inline, otherwise this looks good!

File: salt/modules/runit.py (hunk 1 above)
Comment (markkimsal) [inline salt/modules/runit.py]: I can remove them and re-issue a pull.

**Comment by basepi** [comment]
Now that I'm looking at this closer, I think we actually need the `__virtual__` if you want this function to be able to be used by the service states.

Additionally, there are some lint errors that need taken care of: https://jenkins.saltstack.com/job/salt-pr-lint-n/12767/violations/

**Comment by markkimsal** [comment]
I don't think a lot of process supervisors are used to manage all OS services, and I couldn't get the desired behavior by adding in __virtual__ and __virtualname__ functionality to the runit module.

Currently, if I define a service to be running in a state file I cannot use the "enabled" flag - which is desirable for most other service providers - because runit doesn't have this functionality defined.  This patch simply defines a runit compatible way of enabling and disabling services so that I can properly set the -provider: runit value in my states and not have the state break if I choose to use - enabled: True

Salt in general doesn't have a fall back mechanism to blend multiple providers into one virtual unit so that if one provider fails to execute the desired command another provider can try to execute it.  Promoting runit to a full service provider would not allow me to interact from the CLI with existing sysv services with "service.avaiable" and other calls.  Currently, I can choose to interact with sysv services and supervisored services by using the appropriate module in the CLI - either service.reload or runit.reload.

Also, the order of selecting the proper virtual provider seems to be dependent on if you're using salt, salt-call --local or salt-ssh.  The lazy loading doesn't seem to be deterministic.

Given all of the above, I don't think it is well suited to promote the runit module to the status of full virtual service provider at this point.

**Comment by basepi** [comment]
Perfect, thanks for the explanation.

**Comment by cachedout** [comment]
@markkimsal Looks like we do have some failing tests here. Could you take a look? Let me know if I can help at all. Thanks!

File: salt/modules/runit.py [hunk 2]
```diff
@@ -164,6 +164,85 @@ def available(name):
     return name in get_all()
 
 
+def enabled(name, **kwargs):
```
Comment (cachedout) [inline salt/modules/runit.py]: We prefer that these `**kwargs` be removed. The interface definitions for execution modules should be as well-defined as possible. This also helps with our routines which attempt to parse the argspecs in the case of errors.

**Comment by markkimsal** [comment]
I can't seem to get the tests to pass.  I'm running this on a few or my systems and enable/disable/enabled/disabled all work fine.  It seems to be an issue with setting up Mocks properly.

      def test_enabled(self):
        '''
        Test if it returns ``True`` if the specified service is available,
        otherwise returns ``False``.
        '''
        with patch.object(os, 'listdir',
                          MagicMock(return_value=['run', 'supervise'])):
            mock_mode = MagicMock(return_value='0700')
            with patch.dict(runit.__salt__, {'file.get_mode': mock_mode}):
                self.assertTrue(runit.enabled('foo'))
`

I've tried mock_mode as a simple string '0700' and as a MagicMock return value.  I need to be able to simulate a listdir on /etc/service/foo  that retuns a file 'run' and a directory 'supervise' (just 2 strings) and then mock file.get_mode to return the string 0700.  The module's enabled function should use atoi with base=8 to interpret the executable bit.

If there's something obviously wrong with this test and the use of Mocks, let me know.  Otherwise I'll have to setup a full salt testing environment and try to figure out how to mock the value '0700'.

**Comment by cachedout** [comment]
I've fixed the failing tests for you in #30883. When that is added, this will automatically be closed and your commits will be merged into Salt. Thanks very much @markkimsal !

**Comment by markkimsal** [comment]
Thanks @cachedout for the unit test fix.  I see what I missed now.

**Pull request closed without merging**